### ISSUE TITLE
Not compatible with Proton's fsync (wrong futex opcode)

### ISSUE THREAD

**Issue by imaami** [zen-kernel/zen-kernel#152]
Weird that no one's noticed this yet, but the current zen-kernel uses a branch of futex-wait-multiple with a different `FUTEX_WAIT_MULTIPLE` opcode definition than what Proton's Wine expects. In practice this means that when running a zen or liquorix kernel, Proton's runtime fsync check will fail, and the fsync feature will never be used.

The offending definition is on [line 24 in include/uapi/linux/futex.h](https://github.com/zen-kernel/zen-kernel/blob/d7d1b3d3f9b9fdd2d6da2e69f30f458127131f04/include/uapi/linux/futex.h#L24). Based on the commit message it seems to have been lifted from the [futex-wait-multiple-RFC branch of krisman/linux](https://gitlab.collabora.com/krisman/linux/tree/futex-wait-multiple-RFC).

The problem is that there are several branches, and Valve's Wine patches rely on the value of `FUTEX_WAIT_MULTIPLE` being 31 instead of 13, as is the case in [futex-wait-multiple-master](https://gitlab.collabora.com/krisman/linux/tree/futex-wait-multiple-master) thanks to [this commit](https://gitlab.collabora.com/krisman/linux/commit/368eb7d8c86bd2a5db200567e82c477efc1a9502). See the crucial difference:

* https://gitlab.collabora.com/krisman/linux/blob/futex-wait-multiple-RFC/include/uapi/linux/futex.h#L24
* https://gitlab.collabora.com/krisman/linux/blob/futex-wait-multiple-master/include/uapi/linux/futex.h#L24

Here are the instances where Proton's Wine uses fsync opcode 31 specifically in its `futex_wait_multiple()` wrappers:

* [dls/ntdll/fsync.c](https://github.com/ValveSoftware/wine/blob/49dabdcb065161e8d5d3d9b76ca0a514fb0bbc32/dlls/ntdll/fsync.c#L79)
* [server/fsync.c](https://github.com/ValveSoftware/wine/blob/49dabdcb065161e8d5d3d9b76ca0a514fb0bbc32/server/fsync.c#L64)

Sorry for dragging my feet with this bug report. I noticed this a few weeks ago, but work and family have kept me passive and hoping that someone else would notice this, too.

# TL;DR

Cherry pick [commit 368eb7d8c from krisman/linux](https://gitlab.collabora.com/krisman/linux/commit/368eb7d8c86bd2a5db200567e82c477efc1a9502) onto all of the things.

**Comment by damentz** [comment]
I cherry-picked the commit you referenced to 5.2/futex-backports and 5.3/futex-backports:

https://github.com/zen-kernel/zen-kernel/commit/56c847cd3c185d5766eaa1a70cd93f95ba6450ae
https://github.com/zen-kernel/zen-kernel/commit/414768d240ad2991c88e20d54c1453e1ca179485

To be honest, it looks to me like Valve typo'd in opcode 31, since you can see from the original patch that the next consecutive number is 13.  But you know how it is, you can easily immortalize a typo that requires cooperation between two major code bases when one of them is out of your control.

If that's the case, I'm sure the person who introduced 31 instead of 13 as the opcode face palmed for quite a while.

**Comment by imaami** [comment]
> I cherry-picked the commit you referenced to 5.2/futex-backports and 5.3/futex-backports:
> 
> [56c847c](https://github.com/zen-kernel/zen-kernel/commit/56c847cd3c185d5766eaa1a70cd93f95ba6450ae)
> [414768d](https://github.com/zen-kernel/zen-kernel/commit/414768d240ad2991c88e20d54c1453e1ca179485)
> 
> To be honest, it looks to me like Valve typo'd in opcode 31, since you can see from the original patch that the next consecutive number is 13. But you know how it is, you can easily immortalize a typo that requires cooperation between two major code bases when one of them is out of your control.
> 
> If that's the case, I'm sure the person who introduced 31 instead of 13 as the opcode face palmed for quite a while.

Yeah, it's of course possible. :) It could also have been a "temporary" choice for testing before hopefully getting merged upstream at some point. Either way we're stuck with it now for the time being.

**Comment by tuxutku** [comment]
So from my testing fsync still doesn't work in `5.3.0-15.1-liquorix-amd64`. From a game i have tested it says: `esync: up and running.` and there is no mention of `fsync`
